hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -1991,10 +2017,16 @@ def update_shrink_target(self, new_target):
             new = new_target.buffer
             assert sort_key(new) < sort_key(current)
             self.shrinks += 1
-            if new_target.blocks != self.shrink_target.blocks:
+            if (
+                len(new_target.blocks) != len(self.shrink_target.blocks) or
+                list(new_target.each_block_bounds()) !=
```

Comment:
FWIW attrs gives us good equality methods, so it might make sense to just do `new_target.blocks != self.shrink_target.blocks`. That's *technically* a functionality change though (there are some cases in which this would now trigger that wouldn't previously, but I think those are all fine), so up to you if you want to do it. Also we'd need to take care to then exclude the reference to the enclosing example from equality if we do ahead with the thing from my other comment.

Reply:
I think the presence of `all_zero` would make this a net loss.

We don't necessarily want to throw away change-tracking information just because a block was reduced to zeros.